foursight_core/react/api/auth.py

```diff
@@ -0,0 +1,203 @@
+import boto3
+import time
+from .cookie_utils import read_cookie
+from .envs import Envs
+from .jwt_utils import jwt_decode, jwt_encode
+
+
+class Auth:
+
+    def __init__(self, auth0_client_id: str, auth0_secret: str, envs: Envs):
+        self.auth0_client_id = auth0_client_id
+        self.auth0_secret = auth0_secret
+        self.envs = envs
+
+    cache_aws_credentials = {}
+
+    def authorize(self, request: dict, env: str) -> dict:
+        """
+        Verifies that the given request is authenticated AND authorized, based on the authtoken
+        cookie (a JWT-signed-encoded value) in the given request. If so, returns the verified and
+        decoded authtoken as a dictionary. If not, returns a dictionary indicating not authorized
+        and/or not authenticated, and the basic info contained in the authtoken.
+        """
+        try:
+
+            # Read the authtoken cookie.
+
+            authtoken = read_cookie(request, "authtoken")
+            if not authtoken:
+                return self._create_not_authenticated_response(request, "no-authtoken")
+
+            # Decode the authtoken cookie.
+
+            authtoken_decoded = self.decode_authtoken(authtoken)
+            if not authtoken_decoded:
+                return self._create_not_authenticated_response(request, "invalid-authtoken")
+
+            # Sanity check the decoded authtoken.
+
+            if authtoken_decoded["authorized"] is not True or authtoken_decoded["authenticated"] is not True:
+                return self._create_not_authenticated_response(request, "invalid-authtoken-auth", authtoken_decoded)
+
+            if self.auth0_client_id != authtoken_decoded["aud"]:
```

Comment:
The Auth0 "aud" field is the "audience" claim, and identifies the "recipients" of the JWT according to the doc, by which is meant the *application* itself, being the recipient. I think this can be just about anything, but the way it was coded in Foursight as I found it was to use the Auth0 client IID as this value, so I did the same for the React implementation. A JWT *must* have an "aud" field, otherwise an exception will be raised (by the Python jwt library) on decode. I'll add a (JWT_AUDIENCE_PROPERTY_NAME) constant for this value.